src/module/rules/rule-element/substitute-roll.ts

```diff
@@ -1,70 +1,54 @@
-import { sluggify, tupleHasValue } from "@util";
-import { RuleElementOptions, RuleElementPF2e } from "./base.ts";
-import { RuleElementSource } from "./index.ts";
+import { sluggify } from "@util";
+import { RuleElementPF2e } from "./base.ts";
+import { RuleElementSchema } from "./index.ts";
+import type { BooleanField, StringField } from "types/foundry/common/data/fields.d.ts";
+import { ResolvableValueField } from "./data.ts";
 
 /** Substitute a pre-determined result for a check's D20 roll */
-class SubstituteRollRuleElement extends RuleElementPF2e {
-    selector = "check";
-
-    /** Whether this substitution is required for matching rolls */
-    required = false;
-
-    /** The effect type of this substitution */
-    effectType: "fortune" | "misfortune";
-
-    constructor(data: SubstituteRollSource, options: RuleElementOptions) {
-        data.slug ??= options.parent.slug ?? sluggify(options.parent.name);
-
-        super(data, options);
-
-        if (this.#isValid(data)) {
-            this.selector = data.selector;
-            this.required = data.required ?? false;
-        }
-
-        this.effectType = tupleHasValue(["fortune", "misfortune"] as const, data.effectType)
-            ? data.effectType
-            : "fortune";
-    }
-
-    #isValid(data: SubstituteRollSource): data is SubstituteRollData {
-        return typeof data.selector === "string" && (!("required" in data) || typeof data.required === "boolean");
+class SubstituteRollRuleElement extends RuleElementPF2e<SubstituteRollSchema> {
+    static override defineSchema(): SubstituteRollSchema {
+        const { fields } = foundry.data;
+        return {
+            ...super.defineSchema(),
+            selector: new fields.StringField({ required: true, blank: false, initial: "check" }),
+            value: new ResolvableValueField({ required: true, nullable: false }),
+            required: new fields.BooleanField({ required: false, nullable: false, initial: false }),
+            effectType: new fields.StringField({
+                required: false,
```

Comment:
```suggestion
                required: true,
```

Reply:
There are many existing REs that have omitted the `effectType` because it defaults to `fortune`.

Reply:
Yeah, but it has an `initial` just below, so it's ok!